File: session07/LinkedList.java
public class LinkedList {
    private static class Node {
	int val;
	Node next;
	public Node(int v, Node n) { val = v; next = n; }
    }
    private Node head;
    private Node tail;
    public LinkedList() {
	head = tail = null;
    }
    /* only in C++, but for the test, you need to show pseudocode
    public void destructor() {
	for (Node p = head; p != null; p = p.next)
	    delete p;
    }
     */

    public void addStart(int v) {
	head = new Node(v, head);
	if (tail == null)
	    tail = head;
    }

    public void addEnd(int v) {
	if (tail == null) {
	    head = tail = new Node(v, null);
	    return;
	}
	tail.next = new Node(v, null);
    }

    public void insert(int pos, int val) {
	if (head == null) {
	    head = tail = new Node(val, null);
	    return;
	}
	Node p = head;
	for (; pos > 0; pos--)
	    p = p.next;
	p.next = new Node(val, p.next);
    }

    public boolean isEmpty() {
	return head == null;
    }
    
    public void removeStart() {
	if (head == tail) {
	    head = tail = null;
	    return;
	}
	head = head.next;
    }

    public void removeEnd() {
	// left as an exercise!!!
    }

    public static class Iterator { // LinkedList.Iterator
	private Node current;
	public Iterator(LinkedList list) {
	    current = list.head;
	}
	public boolean done() {
	    return current == null;
	}

	public void next() {
	    current = current.next;
	}
	public int getValue() {
	    return current.val;
	}
    }

    public Iterator  begin() {
	return new Iterator(this);
    }
    
    public static void main(String[] args) {
	LinkedList a = new LinkedList(); // O(1)
	// head = null, tail = null
	a.addStart(5); //O(1)
	a.addEnd(3);   //O(1)
	final int n = 10;
	for (int i = 0; i < n; i++)
	    a.insert(i, i); // O(n*n)

	for (int i = 0; i < n; i++) // O(n*n)
	    System.out.print(a.get(i) + " ");
	System.out.println();

	for (Node p = a.head; p != null; p = p.next) {
	    System.out.print(p.val + " ");
	}

	for (LinkedList.Iterator i = a.begin(); !i.done(); i.next())
	    System.out.println(i.getValue());
	



	
    }

}

File: session07/LinkedList.cc
class LinkedList {
private:
  class Node {
  public:
    int val;
    Node* next;
    Node(int v, Node* n) : val(v), next(n) {}
  };
  Node* head;
  Node* tail;
public:
  LinkedList(): head(nullptr), tail(nullptr) {}
  ~LinkedList() {
    for (Node* p = head; p != nullptr; ) {
      Node* temp = p;
      p = p->next;
      delete temp;
    }
  }
  LinkedList(const LinkedList& orig) {
    

  }
  LinkedList& operator =(constLinkedList& orig) {}
  void addFirst(int v) {
    Node* temp = new Node();
    temp->val = v;
    temp->next = head;
    head = temp;
    //    head = new Node(v, head);
    if (tail == nullptr)
      tail = head;
  }
  void addLast(int v) {
#if 0
    Node* temp = new Node();
    temp->val = v;
    temp->next = nullptr;
    tail->next = temp;
#endif
    if (tail == nullptr) {
      head = tail = new Node(v, nullptr);
      return;
    }
    tail->next = new Node(v, nullptr);
    

  }
  bool isEmpty() const {
    return head == nullptr;
  }
  
  void removeFirst() {
    Node* temp = head;
    head = head->next;
    delete temp;
  }
  
  void removeLast() {
    if (head == tail) {
      delete head;
      head = tail = nullptr;
      return;
    }
    Node* p;
    for (p = head; p->next != tail; p = p->next)
      ;
    p->next = nullptr;
    delete tail;
    tail = p;
  }
  void insert(int pos, int val) {
    Node* p = head;
    for (int i = 0; i < pos; i++)
      p = p->next;
    Node* temp = new Node(val, p->next);
    p->next = temp;
  }

  
  class Iterator {
  private:
    Node* current;
  public:
    Iterator(const LinkedList& list) {
      current = list.head;
    }
    bool operator !() const {
      return current != nullptr;
    }
    // this is ++i
    void operator ++() {
      current = current->next;
    }
    
    // thisoperator is i++
    void operator ++(char*) {
      current = current->next;
    }
    int operator *() const {
      return current->val;
    }
  };
};

int main() {
  LinkedList a;
  a.addFirst(3);
  a.addLast(4);
  a.insert(0, 1); // 3 1 4


  for (Node* p = a.head; p != nullptr; p = p->next) {
    cout << p->val;
  }

  for (LinkedList::Iterator i = a; !i ; ++i)
    cout << *i << " ";
  cout << '\n';

  
  
}
